pkg/controllers/pod_controller.go

```diff
@@ -211,16 +211,24 @@ func (r *PodReconciler) assignGatedPodsToTopology(ctx context.Context, leaderPod
 
 func (r *PodReconciler) leaderPodForFollower(ctx context.Context, pod *corev1.Pod) (*corev1.Pod, error) {
 	log := ctrl.LoggerFrom(ctx)
+	labelSelector := labels.SelectorFromSet(map[string]string{
+		jobset.JobKey: pod.Labels[jobset.JobKey],
+	})
 
-	// Get leader pod if it exists.
-	var leaderPod corev1.Pod
-	leaderName := leaderPodName(pod)
-	log.Info(fmt.Sprintf("leader pod is %q for follower pod %q", leaderName, pod.Name))
-	if err := r.Get(ctx, types.NamespacedName{Name: leaderName, Namespace: pod.Namespace}, &leaderPod); err != nil {
-		log.Info(fmt.Sprintf("leader pod %q does not exist", leaderName))
+	var podList corev1.PodList
+	if err := r.List(ctx, &podList, client.InNamespace(pod.Namespace), &client.ListOptions{LabelSelector: labelSelector}); err != nil {
```

Comment:
we should list once outside the function (we do a list in assignGatedPodsToTopology) and use that to iterate over the rest of the follower pods

Reply:
now that we got the functionality working, I did a refactor to remove unnecessary RPCs, simplify logic, reduce code duplication. let me know if you have any additional thoughts